attestation-agent/attester/src/tsm_report/mod.rs

```diff
@@ -0,0 +1,252 @@
+            e
+        })?;
+
+        Ok(Self { path })
+    }
+    pub fn attestation_report(
+        &self,
+        provider_data: TsmReportData,
+    ) -> Result<Vec<u8>, TsmReportError> {
+        let report_path = self.path.as_path();
+
+        let report_data = match provider_data {
+            TsmReportData::Tdx(inblob) => inblob,
+            TsmReportData::Sev(privlevel, inblob) => {
+                // TODO: untested
+                std::fs::write(report_path.join("privlevel"), vec![privlevel])
+                    .map_err(|e| TsmReportError::Access("privlevel", e))?;
+                inblob
+            }
+        };
+
+        if report_data.is_empty() {
+            return Err(TsmReportError::InblobLen);
+        }
+
+        std::fs::write(report_path.join("inblob"), report_data)
+            .map_err(|e| TsmReportError::Access("inblob", e))?;
+
+        let q = std::fs::read(report_path.join("outblob"))
+            .map_err(|e| TsmReportError::Access("outblob", e))?;
+
+        check_inblob_write_race(report_path)?;
+
+        Ok(q)
+    }
+    pub fn supplemental_data(&self) -> Result<Vec<u8>, TsmReportError> {
+        let report_path = self.path.as_path();
+
+        let aux = std::fs::read(report_path.join("auxblob"))
+            .map_err(|e| TsmReportError::Access("auxblob", e))?;
+
+        check_inblob_write_race(report_path)?;
+
+        Ok(aux)
+    }
+}
+
+// check_inblob_write_race checks that the returned outblob/auxblob
+// matches the quote generation request originally triggered when
+// inblob was written by the TsmReportPath instance. It prevents
+// the race condition that someone else could use the same temporary
+// directory to generate a quote.
+fn check_inblob_write_race(report_path: &Path) -> Result<(), TsmReportError> {
+    let g = std::fs::read_to_string(report_path.join("generation"))
+        .map_err(|e| TsmReportError::Access("generation", e))?;
+
+    let generation = g
+        .trim_matches('\n')
+        .to_string()
+        .parse::<u32>()
+        .map_err(TsmReportError::Parse)?;
+
+    if generation > 1 {
+        return Err(TsmReportError::InblobConflict(generation));
+    }
+
+    Ok(())
+}
+
+fn check_tsm_report_provider(
+    report_path: &Path,
+    wanted: TsmReportProvider,
+) -> Result<(), TsmReportError> {
+    let report_provider = std::fs::read_to_string(report_path.join("provider"))
+        .map_err(|e| TsmReportError::Access("provider", e))?;
+
+    match TsmReportProvider::from_str(&report_provider) {
+        Ok(provider) => {
+            if provider == wanted {
+                Ok(())
+            } else {
+                Err(TsmReportError::MissingProvider(wanted, provider))
+            }
+        }
+        Err(e) => Err(TsmReportError::UnknownProvider(e)),
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::fs::write;
+    use tempfile::tempdir;
+
+    #[test]
+    fn test_tsm_report_provider() {
+        struct TestData<'a> {
```

Comment:
can use `rstest` to make the test cases more dry.

Reply:
I can look into it but as it is right now is consistent with other test implementations under attesters.

Reply:
Where did you see that? I found no test impls under attesters using this. btw, we are trying to using rstest in multiple places like https://github.com/confidential-containers/guest-components/blob/main/attestation-agent/deps/crypto/src/native/aes256ctr.rs#L31-L47

But it is ok to leave it as-is. It's not a critical issue.

Reply:
Let me try with `rstest`. It looks promising, IMO. I only checked that attester `Cargo.toml` does not add it which suggests attester tests do not use it. I also had some existing "template" for this so that is why I ended up doing it like this

Reply:
should be a bit better now, thanks!